### ISSUE TITLE
Incorrect BigQuery schema for decimal, timestamp and date schema on save

### ISSUE THREAD

**Issue by ghost** [spotify/spark-bigquery#19]
Currently, BigQuery schema seems to be inferred from the loaded Avro schema on write. Due to BigQuery spark-avro implementation lacking support for Avro spec logical types annotations, DecimalType gets loaded into Bigquery as StringTimestampType is loaded as Integer and DateType is not recognized at all. 

Should BigQuery schema be explicitly specified rather than using the mismatched Avro type?

**Comment by nevillelyh** [comment]
Yes that'll be a nice to have but would require a lot more work, e.g. schema and record conversion between Avro <-> BigQuery. Plus we loose and space and performance gain from using Avro.

This project is not top priority for us right now but we'd appreciate feedbacks and PRs.

**Comment by ghost** [comment]
Would setting TableSchema on JobConfigurationLoad preclude the use of avro source format?

If that were the case, the BQ TableSchema could be built from sql schema, which wouldn't be too bad.

How do your currently handle writing timestamps and dates to BigQuery? Do you just cast on use?

**Comment by nevillelyh** [comment]
Not sure but I suspect it won't. Their [doc](https://cloud.google.com/bigquery/data-formats) on Avro types doesn't mention `TIMESTAMP`.

In our case we mostly use Long for epoc timestamp and string for dates to avoid this conversion problem.

**Comment by yu-iskw** [comment]
According my research, BigQuery can deal with date even though the data type is string. Off course, the date functions in BigQuery works with string date. So how about changing the data type for timestamp from long to string?

https://github.com/spotify/spark-bigquery/blob/master/src/main/scala/com/databricks/spark/avro/SchemaConverters.scala#L227

**Comment by smith-m** [comment]
with JobConfigurationLoad().setSchema(), Bigquery will load data from Avro based on an arbitrary schema. This only works for avro types that Bigquery can convert internally to the BQ schema type.

In other words, we can load data into BQ with more complete schema (TIMESTAMP, DATE, (FLOAT or STRING for Decimal), STRUCT and ARRAY). Given that spark-avro does not follow Avro convention for date and timestamps (in terms of both data types and schema labels), defining BQ schema instead of relying on avro schema may be the way to go. We can still take advantage of avro load performance and compression.

**Comment by markncooper** [comment]
I currently have forked this repo to accomplish exactly this:

https://github.com/markncooper/spark-bigquery-1/commits/master

The code borrows from the Appflyer repo to generate the schema. There were also a couple of other small gotchas to get this to work. Happy to restructure this and create a pull request if there is interest.

**Comment by nevillelyh** [comment]
Yes that'd be great. Appreciate the help.
@richwhitjr can you take a look please?

**Comment by richwhitjr** [comment]
@markncooper @nevillelyh  Sure let me take a look.  Probably will be more likely tomorrow at this point.

**Comment by richwhitjr** [comment]
@markncooper Making a Pull Request would be helpful, that will make it easier to understand the change.

**Comment by vimota** [comment]
Just wanted to update that BigQuery now supports Logical Types during import (if your project is whitelisted or you use the useAvroLogicalTypes loadConfig flag - we will enable by default, but wanted to avoid breaking users)[1].

[1] https://issuetracker.google.com/issues/35905894
